### PR TITLE
Adding last batch policy for image loaders

### PR DESCRIPTION
**Pull request by SundarRajan98**
- Adding last batch policy for image loaders
- Supports FILL, DROP and PARTIAL last batch policies

### CODE REVIEW

File: rocAL/include/loaders/image/image_loader.h [hunk 1]
```diff
@@ -55,6 +55,7 @@ class ImageLoader : public LoaderModule {
     void shut_down() override;
     void feed_external_input(const std::vector<std::string>& input_images_names, const std::vector<unsigned char*>& input_buffer,
                              const std::vector<ROIxywh>& roi_xywh, unsigned int max_width, unsigned int max_height, unsigned int channels, ExternalSourceFileMode mode, bool eos) override;
+    size_t last_batch_padded_size() override;
```
Comment (rrawther) [inline rocAL/include/loaders/image/image_loader.h:58]: why is this needed for external source reader?

File: rocAL/include/pipeline/context.h [hunk 2]
```diff
@@ -27,10 +27,10 @@ THE SOFTWARE.
 #include "pipeline/master_graph.h"
 
 struct Context {
-    explicit Context(size_t batch_size, RocalAffinity affinity, int gpu_id, size_t cpu_thread_count, size_t prefetch_queue_depth, RocalTensorDataType output_tensor_type) : affinity(affinity),
-                                                                                                                                                                            _user_batch_size(batch_size) {
+    explicit Context(size_t batch_size, RocalAffinity affinity, int gpu_id, size_t cpu_thread_count, size_t prefetch_queue_depth, RocalTensorDataType output_tensor_type, RocalBatchPolicy last_batch_policy, bool last_batch_padded) : affinity(affinity),
```
Comment (rrawther) [inline rocAL/include/pipeline/context.h:30]: in the same enum, we can combine those 2 cases so there will be only one extra parameter for the context

File: rocAL/include/pipeline/master_graph.h [hunk 3]
```diff
@@ -102,6 +102,9 @@ class MasterGraph {
     Status run();
     Timing timing();
     RocalMemType mem_type();
+    RocalBatchPolicy last_batch_policy();
+    bool last_batch_padded();
```
Comment (rrawther) [inline rocAL/include/pipeline/master_graph.h:106]: please combine 105 and 106 into a single api call. For e.g if last_batch_padded_size is 0, last_batch_padded is false

File: rocAL/include/readers/image/image_reader.h [hunk 4]
```diff
@@ -80,6 +80,10 @@ struct ReaderConfig {
     void set_frame_step(unsigned step) { _sequence_frame_step = step; }
     void set_frame_stride(unsigned stride) { _sequence_frame_stride = stride; }
     void set_external_filemode(ExternalSourceFileMode mode) { _file_mode = mode; }
+    void set_last_batch_policy(RocalBatchPolicy last_batch_policy, bool last_batch_padded) {
```
Comment (rrawther) [inline rocAL/include/readers/image/image_reader.h:83]: please combine padded cases into the enum RocalBatchPolicy so only one parameter is passed.

File: rocAL/source/api/rocal_api_data_loaders.cpp [hunk 5]
```diff
@@ -192,7 +192,7 @@ rocalJpegFileSourceSingleShard(
         output = context->master_graph->create_loader_output_tensor(info);
         auto cpu_num_threads = context->master_graph->calculate_cpu_num_threads(shard_count);
 
-        context->master_graph->add_node<ImageLoaderSingleShardNode>({}, {output})->init(shard_id, shard_count, cpu_num_threads, source_path, "", StorageType::FILE_SYSTEM, decType, shuffle, loop, context->user_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original);
+        context->master_graph->add_node<ImageLoaderSingleShardNode>({}, {output})->init(shard_id, shard_count, cpu_num_threads, source_path, "", StorageType::FILE_SYSTEM, decType, shuffle, loop, context->user_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original, context->master_graph->last_batch_policy(), context->master_graph->last_batch_padded());
```
Comment (rrawther) [inline rocAL/source/api/rocal_api_data_loaders.cpp:195]: please see my comment before. Make it into one parameter

File: rocAL/source/pipeline/master_graph.cpp [hunk 6]
```diff
@@ -93,31 +93,33 @@ MasterGraph::~MasterGraph() {
...
-                                                                                                                                                                                     _rb_block_if_full_time("Ring Buffer Block IF Full Time") {
+                                                                                                                                                                                                                                                 _rb_block_if_empty_time("Ring Buffer Block IF Empty Time"),
+                                                                                                                                                                                                                                                 _rb_block_if_full_time("Ring Buffer Block IF Full Time"),
+                                                                                                                                                                                                                                                _last_batch_policy(last_batch_policy),
```
Comment (rrawther) [inline rocAL/source/pipeline/master_graph.cpp:121]: same as my comment before

File: rocAL/source/readers/file_source_reader.cpp [hunk 7]
```diff
@@ -78,6 +70,17 @@ Reader::Status FileSourceReader::initialize(ReaderConfig desc) {
 void FileSourceReader::incremenet_read_ptr() {
     _read_counter++;
     _curr_file_idx = (_curr_file_idx + 1) % _file_names.size();
+    if (_last_batch_info.first == RocalBatchPolicy::DROP) {
+        if (_last_batch_info.second == true) {
+            if ((_file_names.size() / _batch_count) == _curr_file_idx)  // To check if it the last batch
```
Comment (rrawther) [inline rocAL/source/readers/file_source_reader.cpp:75]: Not very clear what this is doing

**Review by rrawther — CHANGES_REQUESTED** [review]
@SundarRajan98 : Let us have a short discussion on the parameter passing for this

File: rocAL/include/loaders/image/image_loader.h (hunk 1 above)
Comment (SundarRajan28) [reply rocAL/include/loaders/image/image_loader.h:58]: This is added to the image loader. We haven't added last batch policy to external source reader since the user would be passing the filenames/buffers for each epoch

File: rocAL/include/readers/image/image_reader.h (hunk 4 above)
Comment (SundarRajan28) [reply rocAL/include/readers/image/image_reader.h:83]: RocalLastBatchPolicy is defined in rocal_api_types.h and used in all rocal API functions. RocalBatchPolicy is defined in commons.h and used internally for all loaders and readers

File: rocAL/include/pipeline/context.h (hunk 2 above)
Comment (SundarRajan28) [reply rocAL/include/pipeline/context.h]: Combined both parameters into a pair variable for the constructor - std::pair<RocalBatchPolicy, bool>

File: rocAL/include/pipeline/master_graph.h [hunk 8]
```diff
@@ -82,7 +82,7 @@ class MasterGraph {
                         NO_MORE_DATA = 2,
                         NOT_IMPLEMENTED = 3,
                         INVALID_ARGUMENTS };
-    MasterGraph(size_t batch_size, RocalAffinity affinity, size_t cpu_thread_count, int gpu_id, size_t prefetch_queue_depth, RocalTensorDataType output_tensor_data_type);
+    MasterGraph(size_t batch_size, RocalAffinity affinity, size_t cpu_thread_count, int gpu_id, size_t prefetch_queue_depth, RocalTensorDataType output_tensor_data_type, RocalBatchPolicy last_batch_policy, bool last_batch_padded);
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL/include/pipeline/master_graph.h]: Combined both parameters into a pair variable for the constructor - std::pair<RocalBatchPolicy, bool>

File: rocAL/include/pipeline/master_graph.h (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/include/pipeline/master_graph.h]: Combined last_batch_policy() and last_batch_padded() function to return a std::pair<RocalBatchPolicy, bool>

File: rocAL/source/api/rocal_api_data_loaders.cpp (hunk 5 above)
Comment (SundarRajan28) [reply rocAL/source/api/rocal_api_data_loaders.cpp]: Replaced it with a single function call - context->master_graph->last_batch_policy_info()

File: rocAL/source/pipeline/master_graph.cpp (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/source/pipeline/master_graph.cpp]: Removed the two internal variables and created a std::pair internal variable which is returned in last_batch_policy_info() function

File: rocAL/source/readers/file_source_reader.cpp (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/source/readers/file_source_reader.cpp:76]: Updated the comments for better understanding

File: rocAL/include/api/rocal_api.h [hunk 9]
```diff
@@ -50,9 +50,13 @@ THE SOFTWARE.
...
+ * \param [in] last_batch_padded Whether the last batch provided by rocAL is padded with the last sample or it just wraps up.
+                                 In the conjunction with last_batch_policy it tells if the iterator returning last batch with 
+                                 data only partially filled with data from the current epoch is dropping padding samples or samples from the next epoch.
  * \return A \ref RocalContext - The context for the pipeline
  */
-extern "C" RocalContext ROCAL_API_CALL rocalCreate(size_t batch_size, RocalProcessMode affinity, int gpu_id = 0, size_t cpu_thread_count = 1, size_t prefetch_queue_depth = 3, RocalTensorOutputType output_tensor_data_type = RocalTensorOutputType::ROCAL_FP32);
+extern "C" RocalContext ROCAL_API_CALL rocalCreate(size_t batch_size, RocalProcessMode affinity, int gpu_id = 0, size_t cpu_thread_count = 1, size_t prefetch_queue_depth = 3, RocalTensorOutputType output_tensor_data_type = RocalTensorOutputType::ROCAL_FP32, std::pair<RocalLastBatchPolicy, bool> last_batch_info = {RocalLastBatchPolicy::ROCAL_LAST_BATCH_FILL, true});
```
Comment (rrawther) [inline rocAL/include/api/rocal_api.h:59]: As I mentioned in the meeting, the last_batch stuff shouldn't be in context. It only affects the loading of data and hence it should be a parameter to appropriate reader/s

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 can you resolve the comments? We need to merge this to mark multi-node as complete

File: rocAL/include/api/rocal_api.h (hunk 9 above)
Comment (rrawther) [reply rocAL/include/api/rocal_api.h:59]: Why is this still needed here?

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 10]
```diff
@@ -66,13 +66,17 @@ class Pipeline(object):
...
+                 last_batch_policy=types.LAST_BATCH_FILL, last_batch_padded=True):
+        if last_batch_padded is False:
+            print("last_batch_padded = False is not implemented in rocAL. Setting last_batch_padded to True")
+            last_batch_padded = True
         if (rocal_cpu):
             self._handle = b.rocalCreate(
-                batch_size, types.CPU, device_id, num_threads, prefetch_queue_depth, tensor_dtype)
+                batch_size, types.CPU, device_id, num_threads, prefetch_queue_depth, tensor_dtype, (last_batch_policy, last_batch_padded))
         else:
             self._handle = b.rocalCreate(
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/pipeline.py:78]: I thought we are removing this from the context/pipeline parameters

**Review by rrawther — CHANGES_REQUESTED** [review]
I still see last_batch_policy being send as part of rocAL parameters

File: rocAL_pybind/amd/rocal/pipeline.py (hunk 10 above)
Comment (SundarRajan28) [reply rocAL_pybind/amd/rocal/pipeline.py]: Removed the LastBatchPolicy from context/master_graph/Pipeline parameters and moved it to decoders.py.

File: rocAL/include/api/rocal_api.h (hunk 9 above)
Comment (SundarRajan28) [reply rocAL/include/api/rocal_api.h]: Removed the LastBatch parameters from rocalCreate and Pipeline creation

**Comment by SundarRajan28** [comment]
@rrawther I have addressed the review comments. Removed the last batch parameters from Context and Pipeline and added them to decoders.py. Please check.

**Review by rrawther — APPROVED** [review]

**Review by LakshmiKumar23 — APPROVED** [review]

**Comment by LakshmiKumar23** [comment]
@kiritigowda Can we merge this. Required for multi-node

**Review by kiritigowda — APPROVED** [review]
